compiler/rustc_lexer/src/lib.rs

```diff
@@ -219,13 +219,6 @@ pub fn strip_shebang(input: &str) -> Option<usize> {
     None
 }
 
-/// Parses the first token from the provided input string.
-#[inline]
-pub fn first_token(input: &str) -> Token {
```

Comment:
The main intention behind `first_token` shape of the API was to guarantee that inceremntal re-lexing works correctly. The cursor can secretly store some state which would make restarting the lexing "from the middle" give a wrong result, while with the `first_token` it is impossible. 

So, from purity perspective, I'd love the `first_token` to be the API basis here, and express token iterator on top of that. 

Practically though, I think exposing cursor is fine.